policy/k8s: Refactor and move `ToServices` translation to policy package

gandro opened the pull request:
This PR moves the logic which translates `ToServices` rules to `ToCIDRSet` rules into `pkg/policy/k8s`. In doing so, we also refactor how translation is performed:

Previously, translation was split over multiple packages: Initial policy ingestion (in `pkg/policy`) invoked `k8s.PreprocessRules`, whichpopulated the initial version of the policy with `ToCIDRSet` rules. Service updates were handled incrementally by the service watcher, which invoked a `k8s.RuleTranslator` for each update and allocating CIDR
identities (via IPCache) accordingly.

The new approach centralizes all the logic into `pkg/policy/k8s`. We mirror the approach from the existing `CiliumCIDRGroup` logic which also translates references to `CiliumCIDRGroups` to `ToCIDRSet`s. Because
the new approach modifies the CNP/CCNP before it is added to the policy repository, we no longer perform incremental updates and no longer need manage CIDR identities ourselves. This simplifies the code and allows us to remove one of the last users of the "synchronous IPCache API". 

Note: This PR removes the last users of the synchronous CIDR allocation API. I will remove those functions (`AllocateCIDRs`/`ReleaseCIDRIdentitiesByCIDR`) in a separate PR since this PR is already on the larger side.

As a side-effect of this change, `ToServices` rules for policies ingested via the non-persistent API (i.e. `cilium-dbg policy import`) are no longer supported. We assume that this was not a commonly used feature for two reasons: First, importing non-K8s policies that refer K8s resources (i.e. services) is likely rather rare. And secondly, the previously implementation was slightly broken: Because policies imported via API were not preprocessed, the policy was defunct until one of the
matched services generated an update event.

Review per commit.

Fixes: #30322

gandro committed:
policy/k8s: Remove double validation in CNP update

This commit removes the `updateCiliumNetworkPolicyV2` and uses the renamed
`addCiliumNetworkPolicyV2` (which is now called
`upsertCiliumNetworkPolicyV2`) in its place.

The removed `updateCiliumNetworkPolicyV2` did not provide any value:

- It validated the old version of the CNP object, even though that old
  version was always coming from the `PolicyWatcher.cnpCache`, to which
  CNPs are only added after they previously passed validation. So the
  `oldRuleCpy.Parse()` call could never fail in practise.
- It called `newRuleCpy.Parse()` followed by a call to
  `addCiliumNetworkPolicyV2` which also would then call `Parse()` again,
  thereby performing the validation twice.

The only meaningful logic was the "Modified CiliumNetworkPolicy" debug
log line, which this commit moves to the `onUpsert` call site. The log
line is also simplified, as there is no way where the name or namespace
of the CNP can change (as we're retrying the old object from the
`cnpCache` which is indexed by name and namespace).

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
policy/k8s: Deduplicate CIDR group resolution logic

This commit moves all code which resolves CIDR group references into a
single `resolveCiliumNetworkPolicyRefs` function. This is done in
preparation for the upcoming sequence of commits, which will resolve
services references in the CNP. By having a single point where we
resolve references, we ensure that all different kinds of references
are updated together and consistently.

This commit does not contain any functional changes.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s/service: Introduce service cache notifications

This introduces a new notification mechanism to the K8s service cache.
In contrast to the existing static single-consumer `Events` channel,
this new `Notifications` mechanism allows multiple dynamic subscribers
to be informed when the service cache changes.

Ideally, we would need two different mechanisms and replace the `Events`
consumer with a `Notifications` consumer. The difficulty comes
from the fact that the current `Events` channel is guaranteed to
deliver all events, including events which occurred before the consumer
started listening on the channel. Where as with the new dynamic
`Notifications` mechanisms, we have to deal with consumers that register
late. Guaranteeing delivery of early events in such a scenario much
harder to achieve, i.e. we would have to replay past events based on the
current cache state.

Since replaying past events would significantly complicate the
implementation, this commit opts for a simpler solution and instead
requires subscribers to register early, i.e. pushing the complexity to
consumers of the new `Notifications` mechanisms. The static consumers of
the existing `Events` mechanism are however unaffected.

There is an upcoming CFP to re-implement the ServiceCache on top of
StateDB. StateDB does have a "watch" mechanism which will give late
subscribers a full review of the cache, thus making custom event
mechanisms obsolete. Until then, the approach implemented by this commit
should serve as a stop-gap solution.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s/service: Introduce service cache iterator

This commit introduces a new helper function to the service cache which
allows users outside of the package to search for services based on
arbitrary conditions (such as e.g. matching labels). The helper will be
used in a subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
policy/api: Allow generated ToCIDRSets with other L3 targets

This commit relaxes the CNP/CCNP policy validation to allow `ToCIDRSet`
rules to coexist with other L3 members, such as `ToServices`. This has
no impact on user-submitted policies, since it is not possible for
policies imported via K8s/API to set the `Generated` field. The
`Generated` field is currently only set by the `ToServices`
implementation, which extracts service endpoints and translates them as
`ToCIDRSet` rules. Currently, this is done after CNP validation, however
a subsequent commit will perform that translation step earlier.

For this reason, we need to relax the rule validation to allow policy
rules to contain both a `ToServices` and generated `ToCIDRSet` rules.

This change has no user-facing impact.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
policy/k8s: Reduce number of CNP copies

This commit changes the `CiliumCIDRGroup` translation logic to not
create a copy of the CNP/CCNP before it is translated. Instead, the
passed in parameter is mutated in place.

While it would arguably be nice to have a functional approach where we
have a input parameter and one output result, creating two copies for
every CNP change is expensive. In particular, since an upcoming change
will add additional translation steps and we want to avoid having to
create unnecessary copies for every translation step.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
policy/k8s: Move ToServices translation to policy package

This commit moves the logic which translates `ToServices` rules to
`ToCIDRSet` rules into `pkg/policy/k8s`. In doing so, we also change how
translation is performed:

Previously, translation was split over multiple packages: Initial policy
ingestion (in `pkg/policy`) invoked `k8s.PreprocessRules`, which
populated the initial version of the policy with `ToCIDRSet` rules.
Service updates were handled incrementally by the service watcher, which
invoked a `k8s.RuleTranslator` for each update and allocating CIDR
identities (via IPCache) accordingly.

The new approach centralizes all the logic into `pkg/policy/k8s`. We
mirror the approach from the existing `CiliumCIDRGroup` logic which also
translates references to `CiliumCIDRGroups` to `ToSericeSet`s. Because
the new approach modifies the CNP/CCNP before it is added to the policy
repository, we no longer perform incremental updates and no longer need
manage CIDR identities ourselves. This simplifies the code and allows us
to remove one of the last users of the "synchronous IPCache API".

As a side-effect of this change, `ToServices` rules for policies
ingested via the non-persistent API (i.e. `cilium-dbg policy import`)
are no longer supported. We assume that this was not a commonly used
feature for two reasons: First, importing non-K8s policies that refer
K8s resources (i.e. services) is likely rather rare. And secondly, the
previously implementation was slightly broken: Because policies imported
via API were not preprocessed, the policy was defunct until one of the
matched services generated an update event.

Note that this commit only removes the conflicting parts the old
infrastructure to keep this commit reviewable. The remaining code of the
old infrastructure is removed in a subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s: Remove k8s.RuleTranslator logic

With the previous commit, this logic is no longer needed.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

learnitall commented on pkg/policy/api/rule_validation.go:
```diff
@@ -179,10 +179,22 @@ func (i *IngressRule) sanitize() error {
 }
 
+// countNonGeneratedRules counts the number of CIDRRule items which are not
+// generated, i.e. were directly provided by the user
```
Can you expand this comment to contain more of the context that you included in the commit message? It would be helpful to have that context when reading the code, rather than having to look for it in the commit history.

gandro replied:
Oh yeah, good point!

learnitall commented on pkg/policy/k8s/service.go:
```diff
@@ -0,0 +1,278 @@
+					appendGeneratedCIDR(prefix)
+				}
+			}
+		}
```
The Prefix struct seems to be a bit large, I could see this function using up a fair amount of memory. Rather than passing the struct into `appendGeneratedCIDR`, I think it would make sense to only pass the output of the `String()` method. This would save on memory by storing a smaller key in the `appendedPrefix` map (https://go.dev/play/p/ozWzfrAJhea) and by not passing the struct as a value into the function.

```suggestion
		appendedPrefix := map[string]struct{}{}
		appendGeneratedCIDR := func(prefix string) {
			if _, found := appendedPrefix[prefix]; found {
				return // avoid inserting same endpoint twice
			}

			rule.Egress[i].ToCIDRSet = append(rule.Egress[i].ToCIDRSet, api.CIDRRule{
				Cidr:      api.CIDR(prefix),
				Generated: true,
			})
			appendedPrefix[prefix] = struct{}{}
		}

		for _, toService := range egress.ToServices {
			if eps := m.cachedSelectors[toService.K8sServiceSelector]; eps != nil {
				for _, prefix := range eps {
					appendGeneratedCIDR(prefix.String())
				}
			} else if eps := m.cachedReferences[toService.K8sService]; eps != nil {
				for _, prefix := range eps {
					appendGeneratedCIDR(prefix.String())
				}
			}
		}
```

gandro replied:
Good idea! I didn't consider the fact that we convert to strings anyways, otherwise I would be worried by the additional string allocation (`netip.Prefix` famously avoids allocations). I might run a quick benchmark nonetheless.

gandro replied:
Marking as resolved since I've removed the dedup logic (see https://github.com/cilium/cilium/pull/31062#issuecomment-1980579585)

learnitall reviewed (changes requested):
Amazing! Thank you so much for all of the detail in your commits, it made reviewing this PR a lot easier. I left two suggestions for you, but otherwise LGTM.

squeed commented on pkg/k8s/service_cache.go:
```diff
@@ -97,12 +98,32 @@ type ServiceEvent struct {
+	// notifications are multicast and may be received by multiple subscribers.
+	notifications stream.Observable[ServiceNotification]
+
+	emitEvent func(ServiceEvent)
```
Is there a reason this is overrideable?

gandro replied:
Because it's a value returned by `stream.Multicast[T]`

squeed replied:
`emitNotification` is, but not `emitEvent`. This is extremely minor splitting hairs, but I would expect `emitEvent()` to be a method, not a function pointer.

gandro replied:
Ah, apologies, the code changes a few times, it used to be `emitNotifications` at some point.

 For `emitEvent` the reason for being a function pointer is that was a convenient way to not have to introduce a separate (send-only) `chan<- ServiceEvent` field alongside `emitNotification`. But that's easily changed, as I don't feel strongly about this.

squeed commented on pkg/policy/k8s/service.go:
```diff
@@ -0,0 +1,278 @@
+	es.SyncRequirementsWithLabelSelector()
+	esMatches := es.Matches(labels.Set(svc.Labels))
+	return esMatches &&
+		(sel.Namespace == svcID.Namespace || sel.Namespace == "")
```
minor nit: can save a lot of comparisons by bailing early if namespace doesn't match.

squeed commented on pkg/policy/k8s/service.go:
```diff
@@ -0,0 +1,278 @@
+	// we have to also want to update any CNPs that matched oldSVC
+	checkOldService := oldSVC != nil && !newSVC.DeepEqual(oldSVC)
+
+	for key, cnp := range p.cnpCache {
```
In high-scale ipcache mode, we will be calling this function a **lot**. Specifically, every time a service changes backend IPs.

We should probably maintain a cache of CNPs / CCNPs that reference services. That way we can make this loop as short as possible.

squeed replied:
On second thought, a mapping of service -> CNP / CCNP names is probably the right thing here; we want to be able to cheaply handle updates here.

gandro replied:
Yeah, fair point. We will still have to iterate over all CNP/CCNP with a service label selector if a service label changes, but for everything else (which I expect to be the more common case) a `svc -> cnp`  map makes sense.

squeed replied:
Oh right, we have selectors. In that case, a list of policies with selectors would probably be a good easy optimization.

squeed commented on pkg/policy/k8s/service.go:
```diff
@@ -0,0 +1,278 @@
+	// The selected endpoints are stored in svcEndpoints.
+	p.svcCache.ForEachService(func(svcID k8s.ServiceID, svc *k8s.Service, eps *k8s.Endpoints) bool {
+		if p.isSelectableService(svc) {
+			svcEndpoints.extractMatches(cnp.Spec, svcID, svc, eps)
```
why are there separate `extractMatches(spec)` and `translateRule(spec)` calls?

gandro replied:
That's a fair question. Mainly because `extractMatches` is invoked for every service **and** rule combination, in service random order. `translateRule` is only invoked for every rule once. Having `translate` be a separate step made it easier to have deterministic rule translation and endpoint deduplication.

But I can take a look again at this though. The overall flow has changed a bit since I first came up with the "extract, then translate" logic and maybe having a single method is much simpler now.

squeed replied:
The only coordination between the two is avoiding duplicate CIDRs, right?

gandro replied:
Yes - looking at the code again, the two can most likely be easily merged. I'll give it a shot.

gandro committed:
policy/k8s: Extract helper for generating resource ID

This commit moves the logic to generate an IPCache resource ID into a
separate function. It will be used by additional call sites in a
subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

learnitall commented on pkg/policy/k8s/service.go:
```diff
@@ -0,0 +1,321 @@
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// isSelectableService returns true if the service svc can be selected by a ToServices rule
```
Is this line missing a period?

learnitall commented on pkg/policy/k8s/service.go:
```diff
@@ -0,0 +1,321 @@
+	if !(newService || changedService) {
+		// If the service definition itself has not changed, and it's not the
+		// first time we process this service, we only need to check the
+		// policies which are known the select the old version of the service
```
I think there might be a typo here:

`policies which are known the select the old version` -> `policies which are known to select the old version`?

learnitall commented on pkg/policy/k8s/service.go:
```diff
@@ -0,0 +1,321 @@
+		}
+
+		// svcEndpoints caches the selected endpoints in case they are
+		// referenced than once by this CNP
```
`references than once by this CNP` -> `referenced more than once by this CNP`?

learnitall commented on pkg/policy/k8s/service.go:
```diff
@@ -0,0 +1,321 @@
+}
+
+// clearCNPForService indicates that a policy (referred to by 'key') no longer
+// has selects service svcID via a ToServices rule
```
`has selects service` -> `selects the service`?

learnitall commented on pkg/policy/k8s/service.go:
```diff
@@ -0,0 +1,321 @@
+			logfields.K8sAPIVersion:           cnp.APIVersion,
+			logfields.K8sNamespace:            cnp.Namespace,
+			logfields.ServiceID:               svcID,
+		}).Debug("Service updated or deleted, recalculating CiliumNetworkPolicy rules")
```
Can we protect this debug call with a log level check?  If this function is called when Services are updated, then depending on the amount of churn the cluster faces this debug statement could become expensive. See https://github.com/cilium/cilium/pull/26327 for context. We're currently working on a way to fix this long-term without this kind of workaround, but for right now it's a good idea to add this kind of log-level check for debug statements in potential hot paths.

learnitall commented on pkg/policy/k8s/watcher.go:
```diff
@@ -170,8 +175,18 @@ func (p *PolicyWatcher) watchResources(ctx context.Context) {
 					err = p.onDeleteCIDRGroup(event.Object.Name, k8sAPIGroupCiliumCIDRGroupV2Alpha1)
 				}
 				event.Done(err)
+			case event, ok := <-serviceEvents:
+				if !ok {
+					serviceEvents = nil
+					break
+				}
+
+				switch event.Action {
+				case k8s.UpdateService, k8s.DeleteService:
+					p.onServiceEvent(event)
+				}
```
I think I might be missing something. From my understanding:

* `p.cnpCache` and `p.toServicesPolicies` are updated when `p.onUpsert` is called while handling a CNP event.
* `p.updateToServicesPolicies` is called by `onServiceEvent` to handle service events.
* `p.updateToServicesPolicies` has a dependency on the `p.cnpCache` and `p.toServicesPolicies` being populated (ie https://github.com/cilium/cilium/pull/31062/files#diff-3c6c6fb5f5d167fcde7a54fe6cc62cfd70d9949eaa0cbb78135f2a11e1d87f4bR68).

Doesn't this dependency create a race condition between a service and any policy that selects it? What happens if a service event is processed before the events for its related CNPs are processed and the `p.cnpCache` and `p.toServicesPolicies` maps are populated? The function `p.watchResources` is written to prioritize CNP events over Service events, but couldn't a service event still be consumed before a CNP event?

If my understanding is correct here, then we should only consume service events after we see a CNP (and any other relevant policy type) sync event, to ensure that `p.cnpCache` and `p.toServicesPolicies` are set before `p.updateToServicesPolicies` is called.

gandro replied:
Great feedback. I hope I understand your concerns correctly:

> `p.updateToServicesPolicies` has a dependency on the `p.cnpCache` and `p.toServicesPolicies` being populated

Correct. The idea of `updateToServicesPolicies` is that it figures out which policies are affected by the service event. If `p.cnpCache` and `p.toServicesPolicies` are not populated, no policies will be updated.

> What happens if a service event is processed before the events for its related CNPs are processed

In such a case, yes the service event is ignored by `updateToServicesPolicies` in the policy subsystem, since it isn't selected by any CNP. But the service subsystem (which populates the service cache, outside of the code changed in this PR) will still keep track of the service in the service cache (i.e. `p.svcCache`). That means, that once the (C)CNP event happens, we still call into `p.resolveToServices` [here](https://github.com/cilium/cilium/blob/3540c4867a3b806d07cbe2fbcf00a3cb30f75b02/pkg/policy/k8s/cilium_network_policy.go#L119) which then queries `p.svcCache` to obtain all references to all selected services. Once the (C)CNP event happened, `p.toServicesPolicies` and `p.cnpByServiceID` should both be populated such that the next  `updateToServicesPolicies`  will do the right thing.

---

Note that there there are potential races around the service handing, but those should be benign: We might observe a service notification for a service which has already been removed. That is fine, because the actual business logic of `p.resolveToServices` always uses the service cache as the source of truth, and never the event itself.

If we get a `service add` notification for a service no longer present in the service cache (i.e. it has already been deleted), we will still call `p.resolveToServices` - but we never actually see the (already deleted) service in `p.svcCache`, thus recalculating the policy, but not actually performing any changes (since the set of service in the service cache remains the same). The good news is that we at least do not re-process policies for `service delete` events for services we have never found in the service cache, since `p.cnpByServiceID` will not reference it.

The other variant, observing a `service add` event not _yet_ present in `p.svcCache` should not occur, since the `p.svcCache` will only notify us once it has added the service to the cache. So that means that if a service event happens and we don't see the the service in the service cache, we can always assume it has been deleted.

To summarize, here are the cases:

- `CNP upsert` first, `service add` second: `CNP upsert` will populate `p.toServicesPolicies` to reference the CNP (in `p.onUpsert`). `service add` then know about the CNP in `updateToServicesPolicies` and will invoke `resolveToServices` (via `resolveCiliumNetworkPolicyRefs`) for that CNP.
- `service add or update` first, `CNP upsert` second: `service add or update` is ignored. `CNP upsert` invokes `p.onUpsert`, which populates `p.toServicesPolicies` and then calls `resolveToServices` (via `resolveCiliumNetworkPolicyRefs`), which will find the service in the service cache and populate `p.cnpByServiceID`.
- `CNP upsert` first, `service update` second: This implies that a service was in the service cache, so `p.cnpByServiceID` is populated during the `CNP upsert` handling. `service update` thus will see it in `p.cnpByServiceID`

So I think all cases should be correct. But I might extend the unit tests to cover those to be certain.

---

Minor side-comment:

> The function p.watchResources is written to prioritize CNP events over Service events

I'm not sure I understand this comment. If two cases in a `select` statement are ready, `select` in Go picks one at random: https://go.dev/ref/spec#Select_statements

learnitall replied:
Thank you for explaining this to me, I really appreciate it. Your breakdown of the different race conditions is very helpful. I missed the service update logic in the CNP event handling. 

And thank you for pointing out the random selection of cases in a `select` statement, I thought they were selected top-to-bottom.

learnitall reviewed (changes requested):
Looks good! I have just have a couple more nits and comments for you this round.

squeed commented on pkg/policy/k8s/service.go:
```diff
@@ -0,0 +1,322 @@
+
+		// svcEndpoints caches the selected endpoints in case they are
+		// referenced more than once by this CNP
+		svcEndpoints := newServiceEndpoints(svcID, svc, eps)
```
Do we even need the `svcEndpoints` object any more? It looks like `processRule()` could be just a function now.

gandro replied:
As it says in the comment: It caches endpoints if they are selected by a policy more than once. It's not necessary for correctness.

learnitall reviewed (approved):
Thank you!
